lib/dereference.js

```diff
@@ -28,60 +28,65 @@ function dereference (parser, options) {
  * @param {string} path - The full path of `obj`, possibly with a JSON Pointer in the hash
  * @param {string} pathFromRoot - The path of `obj` from the schema root
  * @param {object[]} parents - An array of the parent objects that have already been dereferenced
+ * @param {object[]} processedObjects - An array of all the objects that have already been processed
+ * @param {object} dereferencedCache - An map of all the dereferenced objects
  * @param {$Refs} $refs
  * @param {$RefParserOptions} options
  * @returns {{value: object, circular: boolean}}
  */
-function crawl (obj, path, pathFromRoot, parents, $refs, options) {
+function crawl (obj, path, pathFromRoot, parents, processedObjects, dereferencedCache, $refs, options) {
   let dereferenced;
   let result = {
     value: obj,
     circular: false
   };
 
-  if (obj && typeof obj === "object" && !ArrayBuffer.isView(obj)) {
-    parents.push(obj);
+  if (options.dereference.circular === "ignore" || processedObjects.indexOf(obj) === -1) {
```

Comment:
Was `options.dereference.circular` added here accidentally?
I don't think this should be placed here, as this will likely prevent all objects from being crawled.
Let's add a test for it too. We don't seem to have it covered.

The if statement could be combined as well, but that's not a must. It's just a nitpick.

Reply:
in the existing unit tests, it was failing without.
In fact in the process you've got, you dereference the object, then you re-reference if it circular === "ignore"
With my cache it was not execing the re-reference code inside the loop, that's why I add this condition

Reply:
Oh, I misread the condition.
My apologies.
`if (options.dereference.circular !== "ignore" && processedObjects.indexOf(obj) === -1) { `
For whatever reason, I thought of it as above ^.